catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/SourceOperations.java

```diff
@@ -327,42 +328,36 @@ private SourceInfoResponse getFanoutSourceInfo(SourceInfoRequest sourceInfoReque
       }
 
       // Fanout will only add one source descriptor with all the contents
-      Set<Source> availableSources =
-          frameworkProperties
-              .getFederatedSources()
-              .values()
-              .stream()
-              .map(source -> frameworkProperties.getSourcePoller().getCachedSource(source))
-              .filter(source -> source != null && source.isAvailable())
-              .collect(Collectors.toSet());
-
-      Set<ContentType> contentTypes =
-          availableSources
-              .stream()
-              .map(source -> frameworkProperties.getSourcePoller().getCachedSource(source))
-              .filter(source -> source.getContentTypes() != null)
-              .map(Source::getContentTypes)
-              .flatMap(Collection::stream)
-              .collect(Collectors.toSet());
-
-      Source localSource = null;
-
-      if (catalog != null) {
-        localSource = frameworkProperties.getSourcePoller().getCachedSource(catalog);
-        if (localSource != null && localSource.isAvailable()) {
-          availableSources.add(localSource);
-          contentTypes.addAll(localSource.getContentTypes());
-        }
+      // Using a List here instead of a Set because we should not rely on how Sources are compared
+      final List<Source> availableSources =
+          Stream.concat(
+                  frameworkProperties.getFederatedSources().values().stream(),
+                  Stream.of(catalog).filter(Objects::nonNull))
+              .filter(this::isSourceAvailable)
+              .collect(Collectors.toList());
+
+      final Set<ContentType> contentTypes = new HashSet<>();
+      for (Source availableSource : availableSources) {
+        final Set<ContentType> contentTypesForSource =
+            contentTypesCache
+                .getCachedValueForSource(availableSource)
+                .orElseThrow(
+                    () ->
```

Comment:
❓Isn't this going to stop the whole loop as soon as one source is missing from the cache? Is that what we want? Should we log, return an empty set instead and keep going?

Reply:
Updated. I'm not sure if it's worth added a unit test for this case. There is already low coverage for `SourceOperations`.